vnext/Microsoft.ReactNative/Views/TextViewManager.cpp

```diff
@@ -55,7 +68,53 @@ class TextShadowNode final : public ShadowNodeBase {
       Super::AddView(*m_firstChildNode, 0);
       m_firstChildNode = nullptr;
     }
+
     Super::AddView(child, index);
+
+    if (auto run = static_cast<ShadowNodeBase &>(child).GetView().try_as<winrt::Run>()) {
+      if (m_isHighlighted) {
+        AddHighlighter(m_ColorValue, run.Text().size());
+      }
+
+      m_prevCursorEnd += run.Text().size();
+    } else if (auto span = static_cast<ShadowNodeBase &>(child).GetView().try_as<winrt::Span>()) {
+      AddNestedTextHighlighter(
+          m_isHighlighted, m_ColorValue, span, static_cast<VirtualTextShadowNode &>(child).m_highlightData.get());
+    }
+  }
+
+  void AddNestedTextHighlighter(
+      bool isParentHighlighted,
+      folly::dynamic parentColor,
+      winrt::Span &span,
+      VirtualTextShadowNode::HighlightData *highData) {
+    if (!highData->isHighlighted && isParentHighlighted) {
+      highData->isHighlighted = true;
+      highData->color = parentColor;
+    }
+
+    for (const auto &el : span.Inlines()) {
+      if (auto run = el.try_as<winrt::Run>()) {
+        if (highData->isHighlighted) {
+          AddHighlighter(highData->color, run.Text().size());
+        }
+
+        m_prevCursorEnd += run.Text().size();
+      } else if (auto spanChild = el.try_as<winrt::Span>()) {
+        AddNestedTextHighlighter(
+            highData->isHighlighted, highData->color, spanChild, highData->data[highData->spanIdx++].get());
+      }
+    }
+  }
+
+  void AddHighlighter(const folly::dynamic &color, size_t runSize) {
+    auto newHigh = winrt::TextHighlighter{};
+    newHigh.Background(SolidColorBrushFrom(color));
+
+    winrt::TextRange newRange{m_prevCursorEnd, static_cast<int32_t>(runSize)};
+    newHigh.Ranges().Append(newRange);
+
+    this->GetView().as<xaml::Controls::TextBlock>().TextHighlighters().Append(newHigh);
   }
```

Comment:
here we have one TextHighlighter with one range... is there a way to share the text highlighter object instead of creating separate ones?

Reply:
It's something I listed as a possible optimization in the PR description. I could give it more thought, and do it if the time-gain tradeoff is worth it.